File: IntegrationTests/Parser/TypeParsing/TypeParserTests/AssignmentPropagation/AssignmentPropagationSample.cs
// Copyright 2012 rubicon informationstechnologie gmbh
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using InjectionCop.Fragment;

namespace InjectionCop.IntegrationTests.Parser.TypeParsing.TypeParserTests.AssignmentPropagation
{
  class AssignmentPropagationSample: ParserSampleBase
  {
    public void ValidSafenessPropagation()
    {
      string temp = "select * from users";
      RequiresSqlFragment (temp);
    }

    public void InvalidSafenessPropagationParameter([Fragment("SqlFragment")] string temp)
    {
      temp = UnsafeSource(temp);
      RequiresSqlFragment (temp);
    }

    public void ValidSafenessPropagationParameter(string temp)
    {
      DummyMethod (temp);
      temp = SafeSource();
      RequiresSqlFragment (temp);
    }

    public void InvalidSafenessPropagationVariable()
    {
      string temp = SafeSource();
      temp = UnsafeSource(temp);
      RequiresSqlFragment (temp);
    }

    public void ValidSafenessPropagationVariable()
    {
      // ReSharper disable RedundantAssignment
      string temp = UnsafeSource();
      // ReSharper restore RedundantAssignment
      temp = SafeSource();
      RequiresSqlFragment (temp);
    }

    [return: Fragment ("DummyFragment")]
    public string ValidReturnWithIf ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue;
      if (SafeSource() == "Dummy")
      {
        returnValue = "safe";
      }
      else
      {
        returnValue = parameter;
      }
      return returnValue;
    }

    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithIf ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue;
      if (SafeSource() == "Dummy")
      {
        returnValue = "safe";
      }
      else
      {
        returnValue = UnsafeSource();
      }
      return returnValue;
    }

    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithIfFragmentTypeConsidered ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod("dummy");
      }
      else
      {
        returnValue = UnsafeSource();
      }
      return returnValue;
    }

    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithTempVariable ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod("dummy");
      }
      else
      {
        string temp = UnsafeSource();
        DummyMethod (temp);
        returnValue = temp;
      }
      return returnValue;
    }

    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithParameterReset ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod ("dummy");
      }
      else
      {
        parameter = UnsafeSource();
        returnValue = parameter;
      }
      return returnValue;
    }

    [return: Fragment("DummyFragment")]
    private string SafeDummyFragmentSource ()
    {
      return "safe";
    }

    [return: Fragment ("DummyFragment")]
    public string ValidReturnWithParameterReset ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod ("dummy");
      }
      else
      {
        parameter = SafeDummyFragmentSource();
        returnValue = parameter;
      }
      return returnValue;
    }

    [Fragment("DummyFragment")]
    private string _dummyFragmentField = "";
    
    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithFieldReset ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod ("dummy");
      }
      else
      {
        _dummyFragmentField = UnsafeSource();
        returnValue = _dummyFragmentField;
      }
      return returnValue;
    }

    [return: Fragment ("DummyFragment")]
    public string ValidReturnWithFieldReset ([Fragment ("DummyFragment")] string parameter)
    {
      string returnValue = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod ("dummy");
      }
      else
      {
        _dummyFragmentField = SafeDummyFragmentSource();
        returnValue = _dummyFragmentField;
      }
      return returnValue;
    }

    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithField ([Fragment ("DummyFragment")] string parameter)
    {
      _dummyFragmentField = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod("dummy");
      }
      else
      {
        string temp = UnsafeSource();
        DummyMethod (temp);
        _dummyFragmentField = temp;
      }
      return _dummyFragmentField;
    }

    [return: Fragment ("DummyFragment")]
    public string ValidReturnWithField ([Fragment ("DummyFragment")] string parameter)
    {
      _dummyFragmentField = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod("dummy");
      }
      else
      {
        string temp = SafeDummyFragmentSource();
        DummyMethod (temp);
        _dummyFragmentField = temp;
      }
      return _dummyFragmentField;
    }

    /// <summary>
    /// because InjectionCop does not support analyzing while conditions 
    /// this sample is declared invalid although an unsafe value never gets
    /// assigned to fragmentField
    /// </summary>
    /// <param name="parameter"></param>
    /// <returns></returns>
    [return: Fragment ("DummyFragment")]
    public string InvalidReturnWithFieldAndLoops ([Fragment ("DummyFragment")] string parameter)
    {
      _dummyFragmentField = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod("dummy");
      }
      else
      {
        string temp = SafeDummyFragmentSource();
        DummyMethod (temp);
        int i = 0;
        while (i < 10)
        {
          while (i < 5)
          {
            DummyMethod (temp);
            _dummyFragmentField = temp;
            temp = SafeDummyFragmentSource();
            i++;
          }
          temp = UnsafeSource();
          i++;
        }
      }
      return _dummyFragmentField;
    }

    [return: Fragment ("DummyFragment")]
    public string ValidReturnWithFieldAndLoops ([Fragment ("DummyFragment")] string parameter)
    {
      _dummyFragmentField = parameter;
      if (SafeSource() == "Dummy")
      {
        DummyMethod("dummy");
      }
      else
      {
        string temp = SafeDummyFragmentSource();
        DummyMethod (temp);
        int i = 0;
        while (i < 10)
        {
          while (i < 5)
          {
            DummyMethod (temp);
            _dummyFragmentField = temp;
            temp = SafeDummyFragmentSource();
            i++;
          }
          temp = "safe";
          i++;
        }
      }
      return _dummyFragmentField;
    }
  }
}

File: IntegrationTests.VB/Parser/TypeParsing/TypeParserTests/AssignmentPropagation/AssignmentPropagationSample.vb
' Copyright 2013 rubicon informationstechnologie gmbh
'
' Licensed under the Apache License, Version 2.0 (the "License");
' you may not use this file except in compliance with the License.
' You may obtain a copy of the License at
'
' http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS,
' WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
' See the License for the specific language governing permissions and
' limitations under the License.
Imports InjectionCop.Attributes
Imports System

Namespace Parser.TypeParsing.TypeParserTests.AssignmentPropagation
	Friend Class AssignmentPropagationSample
		Inherits ParserSampleBase

		<Fragment("DummyFragment")>
		Private _dummyFragmentField As String = ""

		Public Sub ValidSafenessPropagation()
			Dim temp As String = "select * from users"
			MyBase.RequiresSqlFragment(temp)
		End Sub

		Public Sub InvalidSafenessPropagationParameter(<Fragment("SqlFragment")> temp As String)
			temp = MyBase.UnsafeSource(temp)
			MyBase.RequiresSqlFragment(temp)
		End Sub

		Public Sub ValidSafenessPropagationParameter(temp As String)
			MyBase.DummyMethod(temp)
			temp = MyBase.SafeSource()
			MyBase.RequiresSqlFragment(temp)
		End Sub

		Public Sub InvalidSafenessPropagationVariable()
			Dim temp As String = MyBase.SafeSource()
			temp = MyBase.UnsafeSource(temp)
			MyBase.RequiresSqlFragment(temp)
		End Sub

		Public Sub ValidSafenessPropagationVariable()
			Dim temp As String = MyBase.UnsafeSource()
			temp = MyBase.SafeSource()
			MyBase.RequiresSqlFragment(temp)
		End Sub

		Public Function ValidReturnWithIf(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String
			If MyBase.SafeSource() = "Dummy" Then
				returnValue = "safe"
			Else
				returnValue = parameter
			End If
			Return returnValue
		End Function

		Public Function InvalidReturnWithIf(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String
			If MyBase.SafeSource() = "Dummy" Then
				returnValue = "safe"
			Else
				returnValue = MyBase.UnsafeSource()
			End If
			Return returnValue
		End Function

		Public Function InvalidReturnWithIfFragmentTypeConsidered(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				returnValue = MyBase.UnsafeSource()
			End If
			Return returnValue
		End Function

		Public Function InvalidReturnWithTempVariable(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Dim temp As String = MyBase.UnsafeSource()
				MyBase.DummyMethod(temp)
				returnValue = temp
			End If
			Return returnValue
		End Function

		Public Function InvalidReturnWithParameterReset(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				parameter = MyBase.UnsafeSource()
				returnValue = parameter
			End If
			Return returnValue
		End Function

		Private Function SafeDummyFragmentSource() As<Fragment("DummyFragment")>
		String
			Return"safe"
		End Function

		Public Function ValidReturnWithParameterReset(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				parameter = Me.SafeDummyFragmentSource()
				returnValue = parameter
			End If
			Return returnValue
		End Function

		Public Function InvalidReturnWithFieldReset(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Me._dummyFragmentField = MyBase.UnsafeSource()
				returnValue = Me._dummyFragmentField
			End If
			Return returnValue
		End Function

		Public Function ValidReturnWithFieldReset(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Dim returnValue As String = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Me._dummyFragmentField = Me.SafeDummyFragmentSource()
				returnValue = Me._dummyFragmentField
			End If
			Return returnValue
		End Function

		Public Function InvalidReturnWithField(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Me._dummyFragmentField = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Dim temp As String = MyBase.UnsafeSource()
				MyBase.DummyMethod(temp)
				Me._dummyFragmentField = temp
			End If
			Return Me._dummyFragmentField
		End Function

		Public Function ValidReturnWithField(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Me._dummyFragmentField = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Dim temp As String = Me.SafeDummyFragmentSource()
				MyBase.DummyMethod(temp)
				Me._dummyFragmentField = temp
			End If
			Return Me._dummyFragmentField
		End Function

		Public Function InvalidReturnWithFieldAndLoops(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Me._dummyFragmentField = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Dim temp As String = Me.SafeDummyFragmentSource()
				MyBase.DummyMethod(temp)
        For i As Integer = 0 To 10 - 1
          While i < 5
            MyBase.DummyMethod(temp)
            Me._dummyFragmentField = temp
            temp = Me.SafeDummyFragmentSource()
            i += 1
          End While
          temp = MyBase.UnsafeSource()
        Next
			End If
			Return Me._dummyFragmentField
		End Function

		Public Function ValidReturnWithFieldAndLoops(<Fragment("DummyFragment")> parameter As String) As<Fragment("DummyFragment")>
		String
			Me._dummyFragmentField = parameter
			If MyBase.SafeSource() = "Dummy" Then
				MyBase.DummyMethod("dummy")
			Else
				Dim temp As String = Me.SafeDummyFragmentSource()
				MyBase.DummyMethod(temp)
        For i As Integer = 0 To 10 - 1
          While i < 5
            MyBase.DummyMethod(temp)
            Me._dummyFragmentField = temp
            temp = Me.SafeDummyFragmentSource()
            i += 1
          End While
          temp = "safe"
        Next
			End If
			Return Me._dummyFragmentField
		End Function
	End Class
End Namespace
